### ISSUE TITLE
Proposal for Morphology

### ISSUE THREAD

**Issue by romainbrette** [brian-team/brian2#635]
I finished writing a proposal for a new Morphology class:
https://github.com/brian-team/brian2/wiki/Proposal-for-morphologies

Mainly:
* Basic branch geometry is a sequence of truncated cones.
* Morphology communicates not specific geometrical parameters but only electrically relevant properties to the simulation core (ie area, resistive length etc). The simulation core does not need to know what the actual geometry is.
* Representation (coordinates) is decoupled from electrically relevant properties (eg length could be larger than Euclidian distance between end points).
* For now, multicompartmental somas in SWC files are merged into an isopotential soma.

There are some interesting features that we might want to add later, for example having objects that spread across several branches (for example a multicompartmental soma, of say, the apical dendrite).

**Comment by romainbrette** [comment]
There are a few detail points to discuss. For example, in the current Morphology, properties such as diameter, coordinates etc are initially arrays (one array per branch). But when a SpatialNeuron is created, those are turned into views into a continuous linear array with all compartments. This is not very elegant and it means that a morphology object can be used only once (ie one cannot initialize two neurons with the same morphology object). We might want to do it differently. One option is to duplicate the data. Another option is that in the Morphology object, some properties that are used by the simulation core (area, etc) could be computed instead of stored (and stored in the neuron object as state variables, at construction time).

**Comment by mstimberg** [comment]
Yes, the view thing is something that should change, this is one of the changes from my previously started morphology refactoring that we should apply again. As you said, there are various problems with the approach of changing the `Morphology` object when it is added to the `SpatialNeuron`, and the view approach does not work for standalone anyway.
Another point: Our `Morphology` design necessitates that the objects change after construction (because this is the way you construct a tree), one question is whether we want to allow this still after it was used in a `SpatialNeuron` object. As a concrete example, would the following code to compare the effect of cable diameter be allowed?
```Python
cable = Cylinder(100*um, n=10, diameter=5*um)
neuron_thin = SpatialNeuron(cable, ...)
cable.diameter = 10*um
neuron_thick = SpatialNeuron(cable, ...)
```
I think the easiest would be to allow it and just transform the `Morphology` into a flattened structure when it is used to construct a `SpatialNeuron`.
One more detail is about the use of a `Morphology` for indexing: in the easiest approach, a morphology just returns absolute indices for something like `morph.axon[:50*um]`. Do we care that the user could use the wrong morphology (or a morphology where, say, they changed the lengths after constructing the `SpatialNeuron`) for indexing and therefore get nonsensical indices? In the previous refactoring, I looked into calculating a hash value for the morphology (taking into account all of the structure and all the lengths, etc.) when it is used for constructing the `SpatialNeuron` object, so that you get an error message when you use the wrong `Morphology` object. It complicates things quite a bit, though, and I am not sure it is really worth it (we have the same potential issue when we index synaptic variables for a `StateMonitor`).

**Comment by romainbrette** [comment]
> cable=  Cylinder(100*um,n=10,diameter=5*um)
> neuron_thin=  SpatialNeuron(cable,...)
> cable.diameter=  10*um
> neuron_thick=  SpatialNeuron(cable,...)
>
> I think the easiest would be to allow it and just transform the 
> |Morphology| into a flattened structure when it is used to construct a 
> |SpatialNeuron|.
>
I agree about converting at construction time. The structure is simply 
variables of the group.
But some changes can be problematic. For example, if the length and 
coordinates of a branch is changed (more generally, the end points), it 
makes the tree discontinuous (if the branch has children). We might want 
to think about that.
>
> One more detail is about the use of a |Morphology| for indexing: in 
> the easiest approach, a morphology just returns absolute indices for 
> something like |morph.axon[:50*um]|. Do we care that the user could 
> use the wrong morphology (or a morphology where, say, they changed the 
> lengths after constructing the |SpatialNeuron|) for indexing and 
> therefore get nonsensical indices? In the previous refactoring, I 
> looked into calculating a hash value for the morphology (taking into 
> account all of the structure and all the lengths, etc.) when it is 
> used for constructing the |SpatialNeuron| object, so that you get an 
> error message when you use the wrong |Morphology| object. It 
> complicates things quite a bit, though, and I am not sure it is really 
> worth it (we have the same potential issue when we index synaptic 
> variables for a |StateMonitor|).
>
I'd say it could be a feature for a future version. Not very pressing I 
think.

**Comment by mstimberg** [comment]
> But some changes can be problematic. For example, if the length and coordinates of a branch is changed (more generally, the end points), it makes the tree discontinuous (if the branch has children). We might want to think about that.

I think we decided to not keep these things consistent anyway, right? The morphology/electrical properties are fully defined by the structure, the lengths, etc., we don't care about the coordinates.

**Comment by romainbrette** [comment]
>     But some changes can be problematic. For example, if the length
>     and coordinates of a branch is changed (more generally, the end
>     points), it makes the tree discontinuous (if the branch has
>     children). We might want to think about that.
>
> I think we decided to not keep these things consistent anyway, right? 
> The morphology/electrical properties are fully defined by the 
> structure, the lengths, etc., we don't care about the coordinates.
>
For the simulation it doesn't matter too much I agree.
The issue is more for the morphology object itself. If we allow 
geometrical changes of a branch, then those changes should probably 
trigger a number of recursive changes for the morphology object (eg 
shifting origin of children).

**Comment by mstimberg** [comment]
Ah, now I see what you mean. Actually I am wondering whether this is worth the effort from our side at all -- why not simply forbid all changes to coordinates, lengths, etc. (I guess changing the diameter should be fine)? You can always create a new object if you want to change them (this would also be consistent with the approach that you can't change the segmentation, you'd instead get a new object back).

**Comment by romainbrette** [comment]
Yes, at least for now we could forbid changes in coordinates and length, 
but possibly allow changes in diameters.

Le 26/01/2016 15:54, Marcel Stimberg a écrit :
>
> Ah, now I see what you mean. Actually I am wondering whether this is 
> worth the effort from our side at all -- why not simply forbid all 
> changes to coordinates, lengths, etc. (I guess changing the diameter 
> should be fine)? You can always create a new object if you want to 
> change them (this would also be consistent with the approach that you 
> can't change the segmentation, you'd instead get a new object back).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/brian-team/brian2/issues/635#issuecomment-175057483>.
>

-- 
Twitter: https://twitter.com/RomainBrette
Blog: http://romainbrette.fr/category/blog/

New paper:
Brette R (2015). Philosophy of the spike: rate-based vs. spike-based theories of the brain. Front. Syst. Neurosci. | doi: 10.3389/fnsys.2015.00151

**Comment by mstimberg** [comment]
I am closing this after the merge of #652 , let's open a new issue in case we realize that something is still missing.
